index.js

```
@@ -1,301 +1,319 @@
-    var fname = pathlib.basename( path );
-    var s3 = this._s3;
-    var key = pathlib.join( s3.prefix, date[ 1 ], date[ 2 ], date[ 3 ], fname );
-
-    this.emit( 'upload', path, s3.bucket + '/' + key );
-    s3.putObject({
-        Bucket: s3.bucket,
-        Key: key,
-        Body: fs.createReadStream( path )
-    })
-    .on( 'error', this.emit.bind( this, 'error' ) )
-    .on( 'httpUploadProgress', function ( progress ) {
-        that.emit( 'uploading', path, s3.bucket + '/' + key, progress );
-    })
-    .on( 'success', function () {
-        that.emit( 'uploaded', path, s3.bucket + '/' + key );
-
-        // remove the current path from the queue
-        that._uploadq.shift();
-
-        // if it's not empty, continue to the next one
-        if ( that._uploadq.length ) {
-            that._upload();
-        }
-
-        // remove the uploaded file
-        that._unlink( path )
-    })
-    .send()
-}
-
-RotateStream.prototype._unlink = function ( path, cb ) {
-    var that = this;
-    this.emit( 'delete', path );
-    fs.unlink( path, function ( err ) {
-        if ( err ) {
-            that.emit( 'error', err );
+        if (!stream.compress) {
+            resolve(path)
         } else {
-            that.emit( 'deleted', path );
-            if ( cb ) {
-                cb();
-            }
+            stream.emit('compress', path, pathGzip)
+            fs.createReadStream(path)
+                .pipe(zlib.createGzip())
+                .pipe(fs.createWriteStream(pathGzip))
+                .once('error', (err) => {
+                    stream.emit('error', err)
+                    reject(err)
+                })
+                .once('finish', () => {
+                    stream.emit('compressed', path, pathGzip)
+                    _unlink(path, stream)
+                        .then(() => {
+                            resolve(pathGzip)
+                        })
```

Again - no rejection.

---

I don't understand your comment, if you've meant that the `_unlink` function suppresses errors then this is intentional and now has  a comment stating the reasoning.